src/test/spec/unified_runner/test_file.rs

```diff
@@ -0,0 +1,254 @@
+#![allow(dead_code)]
+
+use semver::{Version, VersionReq};
+use serde::{Deserialize, Deserializer};
+
+use super::TestEvent;
+
+use crate::{
+    bson::{Bson, Document},
+    error::Error,
+    options::{
+        ClientOptions,
+        CollectionOptions,
+        DatabaseOptions,
+        ReadConcern,
+        ReadPreference,
+        SelectionCriteria,
+        WriteConcern,
+    },
+    test::TestClient,
+};
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct TestFile {
+    pub description: String,
+    #[serde(deserialize_with = "deserialize_schema_version")]
+    pub schema_version: Version,
+    pub run_on_requirements: Option<Vec<RunOnRequirement>>,
+    pub allow_multiple_mongoses: Option<bool>,
+    pub create_entities: Option<Vec<Entity>>,
+    pub initial_data: Option<Vec<CollectionData>>,
+    pub tests: Vec<TestCase>,
+}
+
+fn deserialize_schema_version<'de, D>(deserializer: D) -> std::result::Result<Version, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let mut schema_version = String::deserialize(deserializer)?;
+    // if the schema version only contains a major and minor version (e.g. 1.0), append a ".0" to
+    // ensure correct parsing into a semver::Version
+    if schema_version.split('.').count() == 2 {
+        schema_version.push_str(".0");
+    }
+    Version::parse(&schema_version).map_err(|e| serde::de::Error::custom(format!("{}", e)))
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct RunOnRequirement {
+    pub min_server_version: Option<String>,
+    pub max_server_version: Option<String>,
+    pub topology: Option<Vec<String>>,
+}
+
+impl RunOnRequirement {
+    pub fn can_run_on(&self, client: &TestClient) -> bool {
+        if let Some(ref min_version) = self.min_server_version {
+            let req = VersionReq::parse(&format!(">= {}", &min_version)).unwrap();
+            if !req.matches(&client.server_version) {
+                return false;
+            }
+        }
+        if let Some(ref max_version) = self.max_server_version {
+            let req = VersionReq::parse(&format!("<= {}", &max_version)).unwrap();
+            if !req.matches(&client.server_version) {
+                return false;
+            }
+        }
+        if let Some(ref topology) = self.topology {
+            if !topology.contains(&client.topology()) {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub enum Entity {
+    Client(Client),
+    Database(Database),
+    Collection(Collection),
+    Session(Session),
+    Bucket(Bucket),
+    Stream(Stream),
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Client {
+    pub id: String,
+    pub uri_options: Option<ClientOptions>,
```

Comment:
the derived `Deserialize` won't map perfectly here, since this field corresponds to a list of URI options, several of which might combine to single options on `ClientOptions` (e.g. `w` is a URI option, but it is only a component of the `WriteConcern` option).

We may need a custom deserializer for this field which joins all the URI option pairs into a URI itself and parses that or something.